consensus/src/dag/dag_state_sync.rs

```diff
@@ -0,0 +1,202 @@
+
+// TODO: move this to onchain config
+// TODO: temporarily setting DAG_WINDOW to 1 to maintain Shoal safety
+pub const DAG_WINDOW: u64 = 1;
+
+#[async_trait]
+pub trait TUpstreamNotifier: Send {
+    async fn send_epoch_change(&self, proof: EpochChangeProof);
+
+    async fn send_commit_proof(&self, ledger_info: LedgerInfoWithSignatures);
+}
+
+pub(super) struct StateSyncManager {
+    epoch_state: Arc<EpochState>,
+    network: Arc<dyn TDAGNetworkSender>,
+    upstream_notifier: Arc<dyn TUpstreamNotifier>,
+    time_service: TimeService,
+    state_computer: Arc<dyn StateComputer>,
+    storage: Arc<dyn DAGStorage>,
+    dag_store: Arc<RwLock<Dag>>,
+}
+
+impl StateSyncManager {
+    pub fn new(
+        epoch_state: Arc<EpochState>,
+        network: Arc<dyn TDAGNetworkSender>,
+        upstream_notifier: Arc<dyn TUpstreamNotifier>,
+        time_service: TimeService,
+        state_computer: Arc<dyn StateComputer>,
+        storage: Arc<dyn DAGStorage>,
+        dag_store: Arc<RwLock<Dag>>,
+    ) -> Self {
+        Self {
+            epoch_state,
+            network,
+            upstream_notifier,
+            time_service,
+            state_computer,
+            storage,
+            dag_store,
+        }
+    }
+
+    pub async fn sync_to(
+        &self,
+        node: &CertifiedNodeMessage,
+    ) -> anyhow::Result<Option<Arc<RwLock<Dag>>>> {
+        self.sync_to_highest_commit_cert(node.ledger_info()).await;
+        self.try_sync_to_highest_ordered_anchor(node).await
+    }
+
+    /// Fast forward in the decoupled-execution pipeline if the block exists there
+    pub async fn sync_to_highest_commit_cert(&self, ledger_info: &LedgerInfoWithSignatures) {
+        let dag_reader = self.dag_store.read();
+
+        // if the anchor exists between ledger info round and highest ordered round
+        // Note: ledger info round <= highest ordered round
+        if dag_reader
+            .highest_committed_anchor_round()
+            .unwrap_or_default()
+            < ledger_info.commit_info().round()
+            && dag_reader
+                .highest_ordered_anchor_round()
+                .unwrap_or_default()
+                >= ledger_info.commit_info().round()
+        {
+            self.upstream_notifier
+                .send_commit_proof(ledger_info.clone())
+                .await
+        }
+    }
+
+    /// Check if we're far away from this ledger info and need to sync.
+    /// This ensures that the block referred by the ledger info is not in buffer manager.
+    pub fn need_sync_for_ledger_info(&self, li: &LedgerInfoWithSignatures) -> bool {
+        let dag_reader = self.dag_store.read();
+        (dag_reader
+            .highest_ordered_anchor_round()
+            .unwrap_or_default()
+            < li.commit_info().round())
+            || dag_reader
+                .highest_committed_anchor_round()
+                .unwrap_or_default()
+                + 2 * DAG_WINDOW
+                < li.commit_info().round()
```

Comment:
The first part means consensus is behind and the second part means execution, right?
If so, maybe add a comment to explain the condition.